### PR TITLE
Add ability to use a mirror repo for downloading NodeJS installers

### PR DESCRIPTION
**Pull request by Stetchy**
This PR adds functionality for a user to enter a mirror repo when configuring the NodeJS plugin to automatically download and install packages from that repo rather than the official NodeJS repo located at https://nodejs.org/dist.

Reasoning behind this functionality is:
 - Some users may not be able to reach the official repo due to their jurisdiction and may need to use a mirror such as Taobao.
 - Some users working in companies may use software such as Artifactory to mirror public repos instead of allowing internet access.
 - It is not possible to use zip files and install global npm packages - this solution is more flexible than adding ability to install global packages with zip installations.
 
 As this is my first time contributing and I couldn't find much documentation on contributing, I need a small bit of help to get this ready.
 
 I've completed the functionality, it's pretty simple, but in terms of testing I am not sure how to reliably test that the url is being set to the mirror repo, when writing a test the installable url is being set as a temp file. Also, when adding it to the JCasCTests it errors out, not expecting the mirrorRepo option to be available in JCasC, not sure why.
 
 Any pointers on the above would be great, I'd love to get this merged in :)

### CODE REVIEW

**Comment by nfalco79** [comment]
If your company policies does not allow exit to the internet and force to allowed mirror (tipically I expect there is a proxy that access or deny based on white/black list) instead of select "Install automatically" you can set the direct link to the archive.
I do not see how to you had resolve the same issue for JDK, Maven or other tools where is not possible change the certified source. Since I do not see this option available in any other tools I would doubt that the mirror ensure the same content and logic or the original.
Actually I'm not for this approach

**Comment by Stetchy** [comment]
Hi @nfalco79, 

Setting the direct link to the archive doesn't allow installation of global npm packages too. 
Also, when setting a direct link to the zip file you need the complete link for each version, whereas doing it this way allows for use of the drop down and auto-creation of the download URL for whatever version the user chooses.

You can see something similar done when, in NPM we can change the registry URL - it is the same premise. http://npm.taobao.org/mirrors/node is a great example of a mirror repo. With these mirrors it is expected to have the same structure as the NodeJS repo, just with a different base URL. This is why I chose to change even the /dist path, as with mirror repos it doesn't usually exist.

**Comment by nfalco79** [comment]
>Setting the direct link to the archive doesn't allow installation of global npm packages too.

I inherited that behaviour. Any run can perfomr npm install <package> -g and I can not avoid the specified package is installed on NodeJS tool folder that could raise concurrecy issues.

Anyway I understand the case, also my company use Artifactory but we did not use it as mirror. I would ask to you:
* to open a feature on https://issues.jenkins-ci.org/ component nodejs-plugin
* support basic authentication for the mirror (user password credential kind only because download is URL based)
* investigate if there are issue with self-signed certificates and in case if is possible allow them

Last two requirements because it's easy that peoples will ask.

I would leave if possible untouch the Install from secured souce nodejs.org and actual configuration, instead add new Installer specific for mirrored repositories with credentials and self-signed options where the mirror field URL is required and a clear description how the URL work and the expected content. Something like this:
immagine

**Comment by Stetchy** [comment]
Hi @nfalco79,

Will do the above.

I opened an issue here https://issues.jenkins.io/browse/JENKINS-64630 regarding adding the functionality to Jenkins Core. 

I can work on the above features, and in the meantime see what the response is to the ticket above and then we can decide whether we want to add the functionality here, or wait for it to come from core.

Does this sound okay to you?

**Comment by nfalco79** [comment]
I agree. I do not know if mirroring is applicable to any installation tools.
Anyway if you want proceed it would be great. Extending MirrorNodeJSInstaller should enogth to obation the same behavior but a different Installer totally compatible with other code. It is also possible include the jelly page of base NodeJSInstaller class to avoid duplications.

**Comment by Stetchy** [comment]
Cool - I've started work on it and it's looking good, just need to look into the self-signed certs. 

I've created the ticket https://issues.jenkins.io/browse/JENKINS-64646 - would you like me to close this PR then reopen when ready with new criteria in the ticket?

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            boolean useMirror = !mirrorUrl.isEmpty();
+            String baseURL = useMirror ? url.replace("https://nodejs.org/dist", mirrorUrl) : url;
+            url = baseURL + relativeDownloadPath;
+            return this;
+        }
+
+    }
+
+    @Extension
+    public static final class DescriptorImpl extends DownloadFromUrlInstaller.DescriptorImpl<MirrorNodeJSInstaller> { // NOSONAR
```
[the remark this answers is not in the document]
Comment (nfalco79) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java:80]: Add check method for mirrorURL like
`public FormValidation doCheckMirrorURL(@CheckForNull @QueryParameter final String mirrorURL)` where if empty than return an form validation error message or check if the site exists and so on...warn about no env variable are accepted

File: src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#
+# The MIT License
+#
+# Copyright (c) 2016, Riain Condon
```
Comment (nfalco79) [inline src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties:4]: 2021

**Review by nfalco79 — CHANGES_REQUESTED** [review]

File: src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+mirrorUrl.title=Mirror URL
+mirrorUrl.description=Enter the URL of a mirror for the public NodeJS downloads repo here. You may need to do this if you use services like Artifactory to mirror public repos.
+ignoreTlsErrors.title=Don't Verify TLS
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties]: Resolved

File: src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties (hunk 2 above)
Comment (Stetchy) [reply src/main/resources/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller/config.properties]: Resolved

File: src/main/resources/jenkins/plugins/nodejs/Messages.properties [hunk 4]
```diff
@@ -50,3 +50,4 @@ ExecutorCacheLocationLocator.displayName=Local to the executor
 DefaultCacheLocationLocator.displayName=Default (~/.npm or %APP_DATA%\\npm-cache)
 InstallerPathResolver.unsupportedOS=Unresolvable nodeJS installer for version={0}, platform={1}
 InstallerPathResolver.unsupportedArch=Unresolvable nodeJS installer for version={0}, cpu={1}, platform={2}
+MirrorNodeJSInstaller.DescriptorImpl.displayName=Install from mirror
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/resources/jenkins/plugins/nodejs/Messages.properties]: Resolved

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java]: Not sure what you mean about the abstract class - I've extended the NodeJS DescriptorImpl, so I don't have to dupe code, not sure making it abstract while extending DownloadFromUrlInstaller will prevent dupes

[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java]: Added some formvalidation as a start

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        public MirrorNodeJSInstallable(Installable inst) {
+            super(inst);
+        }
+
+        @Override
+        public NodeSpecificInstallable forNode(Node node, TaskListener log) throws IOException {
+            InstallerPathResolver installerPathResolver = InstallerPathResolver.Factory.findResolverFor(id);
+            String relativeDownloadPath = installerPathResolver.resolvePathFor(id, ToolsUtils.getPlatform(node), ToolsUtils.getCPU(node));
+            boolean useMirror = !mirrorUrl.isEmpty();
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java]: Removed

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class MirrorNodeJSInstaller extends NodeJSInstaller {
+    private String mirrorUrl;
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java]: Resolved acronyms

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    private boolean ignoreTlsErrors;
+
+    @DataBoundConstructor
+    public MirrorNodeJSInstaller(String id, String npmPackages, long npmPackagesRefreshHours, String mirrorUrl) {
+        super(id, npmPackages, npmPackagesRefreshHours);
+        this.mirrorUrl = mirrorUrl;
+    }
+
+    public MirrorNodeJSInstaller(String id, String npmPackages, long npmPackagesRefreshHours, boolean force32Bit, String mirrorUrl) {
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java]: Resolved

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
+package jenkins.plugins.nodejs.tools;
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java:24]: Resolved

File: README.md [hunk 9]
```diff
@@ -29,13 +29,15 @@ Plugins Update Center.
...
 -   Relocate npm cache folder using pre defined streategies
+-   Allow use of a mirror repo for downloading and installing NodeJS.
 
 ## Usage
 
 1.  After installing the plugin, go to the global jenkins configuration
     panel (JENKINS\_URL/configure or JENKINS\_URL/configureTools if
     using jenkins 2),  
     and add new NodeJS installations.
+    If you cannot use the official NodeJS repo to install, enter your mirror repo in this section. For example, some users may need to use a mirror due to their jurisdiction.
```
[the remark this answers is not in the document]
Comment (Stetchy) [reply README.md]: Resolved

**Comment by Stetchy** [comment]
Resolved most of the comments, sorry it's taking a bit longer than I expected to get everything done. Quite busy and also figuring out how to do certain things but will get it done asap

File: src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            String relativeDownloadPath = installerPathResolver.resolvePathFor(id, ToolsUtils.getPlatform(node), ToolsUtils.getCPU(node));
+            String baseURL = url.replace(PUBLIC_NODEJS_URL, mirrorURL);
+            url = baseURL + relativeDownloadPath;
+            return this;
+        }
+
+    }
+
+    @Extension
+    public static final class DescriptorImpl extends NodeJSInstaller.DescriptorImpl { // NOSONAR
```
Comment (nfalco79) [inline src/main/java/jenkins/plugins/nodejs/tools/MirrorNodeJSInstaller.java:153]: this descriptor must be typed,. What I mean in my last comment is extends an (new) abstract class parametric in T that have the 3 common methods.

File: src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java [hunk 11]
```diff
@@ -221,7 +220,7 @@ public static boolean areNpmPackagesUpToDate(FilePath expected, String npmPackag
         return marker.exists() && marker.readToString().equals(npmPackages) && System.currentTimeMillis() < marker.lastModified()+ TimeUnit.HOURS.toMillis(npmPackagesRefreshHours);
     }
 
-    private boolean installIfNecessaryMSI(FilePath expected, URL archive, TaskListener listener, String message) throws IOException, InterruptedException {
```
Comment (nfalco79) [inline src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java:224]: the reason of these changes?

**Review by nfalco79 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**